runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,1963 @@
+			if Self::provide_dispute_data(&config, statement_set)? {
+				fresh.push(dispute_target);
+			}
+		}
+
+		Ok(fresh)
+	}
+
+	/// Handle a set of dispute statements corresponding to a single candidate.
+	///
+	/// Fails if the dispute data is invalid. Returns a bool indicating whether the
+	/// dispute is fresh.
+	fn provide_dispute_data(config: &HostConfiguration<T::BlockNumber>, set: DisputeStatementSet)
+		-> Result<bool, DispatchError>
+	{
+		// Dispute statement sets on any dispute which concluded
+		// before this point are to be rejected.
+		let now = <frame_system::Pallet<T>>::block_number();
+		let oldest_accepted = now.saturating_sub(config.dispute_post_conclusion_acceptance_period);
+
+		// Load session info to access validators
+		let session_info = match <session_info::Pallet<T>>::session_info(set.session) {
+			Some(s) => s,
+			None => return Err(Error::<T>::AncientDisputeStatement.into()),
+		};
+
+		let n_validators = session_info.validators.len();
+
+		// Check for ancient.
+		let (fresh, dispute_state) = {
+			if let Some(dispute_state) = <Disputes<T>>::get(&set.session, &set.candidate_hash) {
+				ensure!(
+					dispute_state.concluded_at.as_ref().map_or(true, |c| c >= &oldest_accepted),
+					Error::<T>::AncientDisputeStatement,
+				);
+
+				(false, dispute_state)
+			} else {
+				(
+					true,
+					DisputeState {
+						validators_for: bitvec![BitOrderLsb0, u8; 0; n_validators],
+						validators_against: bitvec![BitOrderLsb0, u8; 0; n_validators],
+						start: now,
+						concluded_at: None,
+					}
+				)
+			}
+		};
+
+		// Check and import all votes.
+		let summary = {
+			let mut importer = DisputeStateImporter::new(dispute_state, now);
+			for (statement, validator_index, signature) in &set.statements {
+				let validator_public = session_info.validators.get(validator_index.0 as usize)
+					.ok_or(Error::<T>::ValidatorIndexOutOfBounds)?;
+
+				// Check signature before importing.
+				check_signature(
+					&validator_public,
+					set.candidate_hash,
+					set.session,
+					statement,
+					signature,
+				).map_err(|()| Error::<T>::InvalidSignature)?;
+
+				let valid = match statement {
+					DisputeStatement::Valid(_) => true,
+					DisputeStatement::Invalid(_) => false,
+				};
+
+				importer.import(*validator_index, valid).map_err(Error::<T>::from)?;
+			}
+
+			importer.finish()
+		};
+
+		// Apply spam slot changes. Bail early if too many occupied.
+		let is_local = <Included<T>>::contains_key(&set.session, &set.candidate_hash);
+		if !is_local {
+			let mut spam_slots: Vec<u32> = SpamSlots::<T>::get(&set.session)
+				.unwrap_or_else(|| vec![0; n_validators]);
+
+			for (validator_index, spam_slot_change) in summary.spam_slot_changes {
+				let spam_slot = spam_slots.get_mut(validator_index.0 as usize)
+					.expect("index is in-bounds, as checked above; qed");
+
+				match spam_slot_change {
+					SpamSlotChange::Inc => {
+						ensure!(
+							*spam_slot < config.dispute_max_spam_slots,
```

Comment:
What should the client do here? Retry and not include the spam statement?

Reply:
The client should not include anything beyond the maximum spam slots. I wrote some docs in the provisioner page of the guide to describe that process